File: ntl.py
import build.ntl as ntl_cpp
import random
from sympy import primefactors
import singleton_timer as st


def GenerateRandomCoprimeLessthanN(n: int):
    assert n >= 1
    num_tries = 0
    while True:
        x = random.randint(1, n)
        num_tries += 1
        if ntl_cpp.Gcd_uint64(x, n) == 1:
            return x, num_tries


def Factorize(n: int):
    assert n >= 2

    factors = primefactors(n)
    return factors


def FindKeyInvModPrime(key: int, mod: int):
    assert key >= 1  # key is always larger or equal to 1
    assert mod >= 1

    # if mod is prime, apply Fermat's Little Theorem
    return ntl_cpp.RepeatedSqrMod_uint64(key, mod - 2, mod)


factors = None
phi = None


# Fastest
def FindKeyInvModNonPrime(key: int, mod: int):
    assert key >= 1  # key is always larger or equal to 1
    assert mod >= 1
    global factors
    global phi  # compute phi only once
    if factors is None:
        factors = primefactors(mod)
        phi = mod
        for factor in factors:
            phi //= factor
        for factor in factors:
            phi *= (factor - 1)
    return ntl_cpp.RepeatedSqrMod_uint64(key, phi - 1, mod)


if __name__ == '__main__':
    pass

File: ntl_cpp/ntl.cpp
#include "ntl.h"
#include <random>

namespace jpyo0803 {
  
std::vector<uint64_t> EratosthenesSieve(uint64_t n) {
  assert(n >= 2); // 'n' needs to be at least 2 

  std::vector<bool> prime_arr(n + 1, true);
  for (uint64_t i = 2; i * i <= n; ++i) {
    if (prime_arr[i]) {
      for (uint64_t j = i * i; j <= n; j += i) {
        prime_arr[j] = false;
      }
    }
  }

  std::vector<uint64_t> primes;
  for (uint64_t i = 2; i <= n; ++i) {
    if (prime_arr[i]) {
      primes.push_back(i);
    }
  }

  return primes;
}

uint64_t FindKeyInvModPrime(uint64_t key, uint64_t mod) {
  assert(key >= 1);
  assert(mod >= 1);
  return RepeatedSqrMod<uint64_t>(key, mod - 2, mod);
}

uint64_t FindKeyInvModNonPrime(uint64_t key, uint64_t mod, const std::vector<uint32_t>& factors) {
    assert(key >= 1);
    assert(mod >= 1);

    static bool first = true;
    static uint64_t phi;
    if (first) {
      phi = mod;
      for (auto factor : factors) {
          phi /= factor;
      }
      for (auto factor : factors) {
          phi *= (factor - 1);
      }
      first = false;
    }
    return RepeatedSqrMod<uint64_t>(key, phi - 1, mod);
}

}